Polyfills/Window/Source/TimeoutDispatcher.cpp

```diff
@@ -0,0 +1,150 @@
+#include "TimeoutDispatcher.h"
+
+#include <assert.h>
+
+namespace Babylon::Polyfills::Internal
+{
+    using TimePoint = std::chrono::time_point<std::chrono::steady_clock, std::chrono::milliseconds>;
+
+    TimePoint Now()
+    {
+        return std::chrono::time_point_cast<std::chrono::milliseconds, std::chrono::steady_clock>(std::chrono::steady_clock::now());
+    }
+
+    struct Timeout
+    {
+        TimeoutId id;
+
+        // Make this non-shared when JsRuntime::Dispatch supports it.
+        std::shared_ptr<Napi::FunctionReference> function;
+
+        TimePoint time;
+
+        Timeout(TimeoutId id, Napi::Function function, TimePoint time)
+            : id{id}
+            , function{std::make_shared<Napi::FunctionReference>(Napi::Persistent(function))}
+            , time{time}
+        {
+        }
+    };
+
+    TimeoutDispatcher::TimeoutDispatcher(Babylon::JsRuntime& runtime)
+        : m_runtime(runtime)
+        , m_thread{&TimeoutDispatcher::ThreadFunction, this}
+    {
+    }
+
+    TimeoutDispatcher::~TimeoutDispatcher()
+    {
+        {
+            std::unique_lock<std::mutex> lk(m_mutex);
+            m_idMap.clear();
+            m_timeMap.clear();
+            m_shutdown = true;
+            m_condVariable.notify_one();
+        }
+
+        if (m_thread.joinable())
+        {
+            m_thread.join();
+        }
+    }
+
+    TimeoutId TimeoutDispatcher::Dispatch(Napi::Function function, std::chrono::milliseconds delay)
+    {
+        if (delay.count() <= 0)
+        {
+            delay = std::chrono::milliseconds{0};
+        }
+
+        std::unique_lock<std::mutex> lk(m_mutex);
+
+        const auto earliestTime = m_timeMap.empty() ? TimePoint::max()
+                                                    : m_timeMap.cbegin()->second->time;
+        const auto id = NextTimeoutId();
+        const auto time = Now() + delay;
+        m_idMap.insert({id, std::make_unique<Timeout>(id, std::move(function), time)});
+        m_timeMap.insert({time, m_idMap[id].get()});
```

Comment:
`insert` returns the iterator where it inserted. You can use that instead of accessing the map again. Something like this:

```suggestion
        const auto it = m_idMap.insert({id, std::make_unique<Timeout>(id, std::move(function), time)});
        m_timeMap.insert({time, it->second.get()});
```